ipn/ipnlocal/local.go

```diff
@@ -1956,13 +1961,44 @@ func (b *LocalBackend) loadStateLocked(key ipn.StateKey, prefs *ipn.Prefs) (err
 	return nil
 }
 
+// setTCPPortsIntercepted populates b.shouldInterceptTCPPortAtomic with an
+// efficient func for ShouldInterceptTCPPort to use, which is called on every
+// incoming packet.
+func (b *LocalBackend) setTCPPortsIntercepted(ports []uint16) {
+	slices.Sort(ports)
+	uniq.ModifySlice(&ports)
+	b.logf("localbackend: handling TCP ports = %v", ports)
+	var f func(uint16) bool
+	switch len(ports) {
+	case 0:
+		f = func(uint16) bool { return false }
+	case 1:
+		f = func(p uint16) bool { return ports[0] == p }
+	case 2:
+		f = func(p uint16) bool { return ports[0] == p || ports[1] == p }
+	case 3:
+		f = func(p uint16) bool { return ports[0] == p || ports[1] == p || ports[2] == p }
+	default:
+		f = func(p uint16) bool {
+			for _, x := range ports {
+				if p == x {
+					return true
+				}
+			}
+			return false
+		}
+	}
+	b.shouldInterceptTCPPortAtomic.Store(f)
+}
+
 // setAtomicValuesFromPrefs populates sshAtomicBool and containsViaIPFuncAtomic
 // from the prefs p, which may be nil.
 func (b *LocalBackend) setAtomicValuesFromPrefs(p ipn.PrefsView) {
 	b.sshAtomicBool.Store(p.Valid() && p.RunSSH() && envknob.CanSSHD())
 
 	if !p.Valid() {
 		b.containsViaIPFuncAtomic.Store(tsaddr.NewContainsIPFunc(nil))
+		b.shouldInterceptTCPPortAtomic.Store(func(uint16) bool { return false })
```

Comment:
Why not `	b.shouldInterceptTCPPortAtomic.Store(nil)`?

Reply:
so caller can blindly call the func. we do this this atomic-func-documented-non-nil thing in a few places

Reply:
but this should be setTCPPortsIntercepted(nil)

Reply:
That's what I meant and I apparently typo'd :+1: